### PR TITLE
sess: stabilize `-Zterminal-width` as `--diagnostic-width`

### PR DESCRIPTION
**Pull request by davidtwco**
Formerly `-Zterminal-width`, `--terminal-width` allows the user or build
tool to inform rustc of the width of the terminal so that diagnostics
can be truncated.

Pending agreement to stabilize, see tracking issue at #84673.

r? @oli-obk

### CODE REVIEW

**Comment by jyn514** [comment]
```
thread 'main' panicked at 'No option 'terminal-width' defined', /cargo/registry/src/github.com-1ecc6299db9ec823/getopts-0.2.21/src/lib.rs:799:21
```
Looks like you need to add this option for rustdoc too - see https://github.com/rust-lang/rust/blob/2ed6786404be276874fbcc7c3540f3237a87e0f3/src/librustdoc/lib.rs#L238-L659 for where the existing options are defined.

File: src/test/ui/terminal-width/flag-human.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// compile-flags: -Z terminal-width=20
+// compile-flags: --terminal-width=20
```
Comment (jyn514) [inline src/test/ui/terminal-width/flag-human.rs:1]: are there equivalent tests for rustdoc? If not, could you add some in `src/test/rustdoc-ui`?

Comment (davidtwco) [reply src/test/ui/terminal-width/flag-human.rs:1]: Added!

**Review by jyn514 — APPROVED** [review]

**Comment by davidtwco** [comment]
> The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5844209841?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ee0665be0e7b9fc3c78a430744aa97c16f6fc3a1/checks/5844209841/logs)
> 
> Click to see the possible cause of the failure (guessed by this bot)

Spurious failure.

**Review by oli-obk — APPROVED** [review]
r=me once stabilization FCP finishes

**Comment by davidtwco** [comment]
nagisa [left a relevant comment in Zulip](https://rust-lang.zulipchat.com/#narrow/stream/147480-t-compiler.2Fwg-diagnostics/topic/terminal.20width.20stabilization/near/279190364) that might be worth considering, maybe a bit late now since stabilization has already finished.

**Comment by oli-obk** [comment]
we can just do a quick mention in the compiler team triage. I like the `output-width` variant.

**Comment by davidtwco** [comment]
> we can just do a quick mention in the compiler team triage. I like the `output-width` variant.

There was [some discussion in Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-06-16/near/286367549). I've pushed a commit that changes `terminal-width` to `output-width`, we can choose whether we keep it.

**Comment by jyn514** [comment]
I think `output-width` is more vague and harder to understand without reading the documentation. `terminal-width` is not accurate in all cases, but it gives a much better idea of what the flag does, and we can't get perfect precision with two words anyway.

**Comment by davidtwco** [comment]
Updated this to be `--diagnostic-width` rather than `--output-width` as that appeared to be [the consensus on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-06-16/near/286397924).

File: src/librustdoc/lib.rs [hunk 2]
```diff
@@ -462,6 +462,14 @@ fn opts() -> Vec<RustcOptGroup> {
                 "human|json|short",
             )
         }),
+        stable("diagnostic-width", |o| {
```
Comment (davidtwco) [inline src/librustdoc/lib.rs:465]: @rust-lang/rustdoc this would insta-stable for rustdoc, would you prefer I add it as unstable first?

**Comment by jyn514** [comment]
I'd like to bump this again: https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-06-16/near/286383610

> @jyn514: as a general thing, I think it's a bad idea to change features as they're stabilized - if someone has a suggestion to improve it, we should merge that on nightly, give it a week or two, and then stabilize

> @pnkfelix I think this is a super important insight. Seems analogous to the "no new rationale" rule that we've had in place forever...

**Comment by davidtwco** [comment]
@oli-obk do you want to take another look at this before r=you?

**Pull request merged**